instructions/src/lib.rs

```diff
@@ -1,162 +1,1532 @@
-        let opcode = match size {
-            MemSize::Double => 0x79,
-            MemSize::Half => 0x69
-        };
-        Operation::new(self, opcode)
+impl AsRef<Endian> for Endian {
+    fn as_ref(&self) -> &Self {
+        self
     }
+}
 
-    pub fn exit(&'p mut self) -> Operation<'p> {
-        Operation::new(self, 0x95)
+impl<'e> From<&'e Endian> for u8 {
+    fn from(endian: &'e Endian) -> u8 {
+        match *endian {
+            Endian::Little => 0xd4,
+            Endian::Big => 0xdc
+        }
     }
 }
 
-pub struct Operation<'p> {
-    program: &'p mut ProgramCodeBuilder,
-    op: u8,
+pub struct Load {
+    mem_size: MemSize,
+    addressing: Addressing,
+    address_source: AddressSource,
     dst: u8,
     src: u8,
     offset: u16,
-    immediate: u32
+    imm: u32
+}
+
+impl Load {
+    pub fn new(mem_size: MemSize, addressing: Addressing, address_source: AddressSource) -> Self {
+        Load {
+            mem_size: mem_size,
+            addressing: addressing,
+            address_source: address_source,
+            dst: 0x00,
+            src: 0x00,
+            offset: 0x00_00,
+            imm: 0x00_00_00_00
+        }
+    }
+}
+
+impl Instruction for Load {
+    fn opt_code_byte(&self) -> u8 {
+        let addressing: u8 = self.addressing.as_ref().into();
+        let size: u8 = self.mem_size.as_ref().into();
+        let address_source: u8 = self.address_source.as_ref().into();
+        addressing | size | address_source
+    }
+
+    fn dst(&self) -> u8 {
+        self.dst
+    }
+
+    fn src(&self) -> u8 {
+        self.src
+    }
+
+    fn offset(&self) -> u16 {
+        self.offset
+    }
+
+    fn imm(&self) -> u32 {
+        self.imm
+    }
+
+    fn dst_reg(mut self, dst: u8) -> Self {
+        self.dst = dst;
+        self
+    }
+
+    fn src_reg(mut self, src: u8) -> Self {
+        self.src = src;
+        self
+    }
+
+    fn offset_bytes(mut self, offset: u16) -> Self {
+        self.offset = offset;
+        self
+    }
+
+    fn immediate(mut self, imm: u32) -> Self {
+        self.imm = imm;
+        self
+    }
+}
+
+pub enum MemSize {
+    DoubleWord,
+    Byte,
+    HalfWord,
+    Word
+}
+
+impl AsRef<MemSize> for MemSize {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<'ms> From<&'ms MemSize> for u8 {
+    fn from(mem_size: &'ms MemSize) -> u8 {
+        match *mem_size {
+            MemSize::DoubleWord => 0x18,
+            MemSize::Byte => 0x10,
+            MemSize::HalfWord => 0x08,
+            MemSize::Word => 0x00
+        }
+    }
+}
+
+pub enum Addressing {
+    Undef,
+    Abs,
+    Ind
+}
+
+impl AsRef<Addressing> for Addressing {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<'a> From<&'a Addressing> for u8 {
+    fn from(addressing: &'a Addressing) -> u8 {
+        match *addressing {
+            Addressing::Undef => 0x00,
+            Addressing::Abs => 0x20,
+            Addressing::Ind => 0x40
+        }
+    }
+}
```

Comment:
Not sure we have to use a match for extracting values from enums. Apparently you can give each item a _discriminant_ value, and then cast as an integer type, see <https://doc.rust-lang.org/reference.html#enumerations>.